tests/src/Kernel/AccessTest.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_contact_forms\Kernel;
+
+use Drupal\contact\Entity\ContactForm;
+
+/**
+ * Test the contact form access control.
+ */
+class AccessTest extends ContactFormTestBase {
+
+  /**
+   * Tests corporate and default contact form access.
+   */
+  public function testContactFormAccess(): void {
+    /** @var \Drupal\Core\Entity\EntityTypeManager $message_storage */
+    $message_storage = $this->container->get('entity_type.manager');
+    /** @var \Drupal\Core\Path\PathValidator $pathValidator */
+    $pathValidator = \Drupal::service('path.validator');
+
+    // Tests corporate contact form access.
+    $contact_form = ContactForm::create(['id' => 'oe_contact_form']);
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'is_corporate_form', TRUE);
+    $contact_form->setThirdPartySetting('oe_contact_forms', 'allow_canonical_url', FALSE);
+    $contact_form->save();
+
+    // Assert user with "access corporate contact form" permission has access
+    // to contact_message entity.
+    $account1 = $this->createUser(['name' => 'test1'], ['access corporate contact form']);
+    $this->assertTrue($message_storage->getAccessControlHandler('contact_message')->createAccess($contact_form->id(), $account1));
+    // Assert user without "access corporate contact form" permission does not.
+    $account2 = $this->createUser(['name' => 'test2'], ['access site-wide contact form']);
+    $this->assertFalse($message_storage->getAccessControlHandler('contact_message')->createAccess($contact_form->id(), $account2));
+
+    // Assert form without allow_canonical_url is forbidden.
+    $this->setCurrentUser($account2);
+    $url = $pathValidator->getUrlIfValid('contact/' . $contact_form->id());
```

Comment:
why are you using the path validator?

Url::fromRoute('contact form blabla route with parameters')->access()

Reply:
nice one, done